crates/rpc/src/method/subscribe_new_transaction_receipts.rs

```diff
@@ -0,0 +1,832 @@
+                                    if !params.matches(&transaction.variant.sender_address(), TxnFinalityStatus::AcceptedOnL2) {
+                                        continue;
+                                    }
+
+                                    let notification = Notification {
+                                        block_hash: Some(block.block_hash),
+                                        block_number: Some(block.block_number),
+                                        receipt: receipt.clone(),
+                                        transaction: transaction.clone(),
+                                        events: events.clone(),
+                                        finality: TxnFinalityStatus::AcceptedOnL2,
+                                    };
+                                    if tx
+                                        .send(SubscriptionMessage {
+                                            notification,
+                                            block_number: block.block_number,
+                                            subscription_name: SUBSCRIPTION_NAME,
+                                        })
+                                        .await
+                                        .is_err()
+                                    {
+                                        // Close subscription.
+                                        return Ok(());
+                                    }
+                                }
+
+                                // Clear the sent transactions set.
+                                pre_confirmed_sent_txs.clear();
+                            }
+                        }
+                    }
+                }
+                pending_changed = pending_data.changed() => {
+                    if let Err(e) = pending_changed {
+                        tracing::debug!(error=%e, "Pending data channel closed, stopping subscription");
+                        return Ok(());
+                    }
+
+                    let pending = pending_data.borrow_and_update().clone();
+                    let finality_status = pending.block().finality_status();
+
+                    tracing::trace!(block_number=%pending.block_number(), ?finality_status, "Pre-confirmed block update");
+
+                    if pending.block_number() != last_pre_confirmed_block {
+                        last_pre_confirmed_block = pending.block_number();
+                        pre_confirmed_sent_txs.clear();
+                    }
+
+                    for (transaction, (receipt, events)) in pending.transactions().iter().zip(pending.transaction_receipts_and_events().iter()) {
+                        if pre_confirmed_sent_txs.contains(&transaction.hash) {
+                            continue;
+                        }
+
+                        if !params.matches(&transaction.variant.sender_address(), pending.block().finality_status()) {
```

Comment:
I suppose the local `finality_status` could be reused here?

Reply:
Indeed, thanks!